rfc/drafts/indexer-ideas-v1.md

```diff
@@ -0,0 +1,303 @@
+  - E.g. `EntityLinkCell = {"group_entity_link":[(ref1,ref2,ref3),(ref4,ref5),(ref6)]}`
+  - Meaning that all grouped entities in that list should have the indicated relationship, and all separated
+    entities should have the inverse relationship. I.e. is_derivation vs is_not_derivation.
+  - Could be easily converted to all-pairs of links internally.
+
++ Object-localization for video / audio / images?
+  - System identifies bounding-boxes for multiple objects in an image / time ranges for multiple known audio clips
+    in an audio broadcast / time ranges for known video clips.
+  - System allows user to focus on just one of those detected objects in their search / dedupe query.
+  - Relevant to the search / dedupe architecture, because this likely would require for those multiple objects
+    to be indexed separately, along with their location in the media.
+
++ Non-monetary incentives for resource providers? I.e. User donates resources because he likes particular content.
+  - Force resource providers to support all content equally (Freenet).
+  - Allow full control over the content you're supporting (Bittorrent).
+  - Something in the middle (Tor exit nodes)?
+
++ Resource usage considerations?
+  - In non-P2P systems, the equation to optimize for tends to be simple. E.g. maximizing utilization of already-owned hardware,
+    or maximizing per-dollar results.
+  - Costs can be different in P2P systems. Some users can get upset and leave, if all of their resources are
+    immediately maxed out.
+  - Considering that P2P networks tend to have a power-law distribution of resources.
+
++ Allowing user to explicitly describe the type of semantic-similarity he's searching for?
+  - Query interpretation was part of textual search engines since very early on - "do know go", but this isn't done for the major image search engines?
+  - Allow users to directly tell the system what kind of semantic similarity they're searching for?
+    - Some indexing systems don't natively deal with multiple types of semantic similarity, or worse, blur multiple types together into an average that doesn't resemble any meaningful type of semantic similarity.
+    - One type of interface: user inputs image - enters a prose description of the type of similarity he's after?
+      - pictures containing same brands and trademarks as this picture
+      - more pictures in the style of this artist
+      - people with same pose as this picture
+      - same type of coffee mug as this picture
+      - more pictures of from this location
+      - more pictures of this person
+      - mashups containing this picture
+  - Or, we start with a small number of specific types of semantic similarity to start.
+  - Range of ways to support this, some relatively simple.
+
++ Eviction of images from the index?
+
++ Co-location of images that should be stored together, for faster retrieval / bulk deletion / cache promotion or eviction?
+
+
+## V1 SEARCH / DEDUPE API:
```

Comment:
Any ideas on how exactly the real-time search system will be connected to the network? Presumably, P2P tracks the addresses of search indexes, and then a direct connection is established between the searcher and the search system?

The dedupe mode could presumably be done in batches and communicate duplicate decisions via the chain. Or, would it make any sense to consider a lazy just-in-time dedupe? (Probably not.)

Isolated components seems to make sense. In the beginning - because several OTC components are being tied together. In the medium term - as we may want to be able to easily swap out components several times as we iterate. It's a highly active research field. In the long run - when moving the components to separate dedicated hardware could be more efficient.

Could be good reasons to embed everything more tightly, though. If so, interested to hear those ideas.